cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/TransformExecutor.java

```diff
@@ -16,92 +16,125 @@
 
 package co.cask.cdap.etl.common;
 
+import co.cask.cdap.api.metrics.Metrics;
 import co.cask.cdap.etl.api.Destroyable;
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.cdap.etl.api.InvalidEntry;
 import co.cask.cdap.etl.api.Transformation;
-import com.google.common.collect.Lists;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
 
-import java.util.ArrayList;
-import java.util.Collection;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
+
 /**
  * Executes Transforms one iteration at a time, tracking how many records were input into and output from
  * each transform.
  *
  * @param <IN> the type of input object to the first transform
- * @param <OUT> the type of object output by the last transform
+ *
  */
-public class TransformExecutor<IN, OUT> implements Destroyable {
-
-
-  private static final Logger LOG = LoggerFactory.getLogger(TransformExecutor.class);
-
-  private final List<TransformDetail> transforms;
-  private final List<DefaultEmitter> emitters;
+public class TransformExecutor<IN> implements Destroyable {
+
+  private final String start;
+  private final Map<String, List<String>> connectionsMap;
+  private final Map<String, Transformation> trackedTransformation;
+  private final DefaultEmitter<Object> defaultEmitter;
+
+  public TransformExecutor(Map<String, Transformation> transformationMap, Metrics metrics,
+                           Map<String, List<String>> connectionsMap, String start) {
+    this.start = start;
+    this.connectionsMap = connectionsMap;
+    this.trackedTransformation = new HashMap<>();
+    for (Map.Entry<String, Transformation> transformationEntry : transformationMap.entrySet()) {
+      trackedTransformation.put(transformationEntry.getKey(),
+                                new TrackedTransform<>(transformationEntry.getValue(),
+                                                     new DefaultStageMetrics(metrics, transformationEntry.getKey())));
+    }
+    defaultEmitter = new DefaultEmitter<>(metrics);
+  }
 
+  public TransformResponse runOneIteration(IN input) throws Exception {
+    if (trackedTransformation.containsKey(start)) {
+      executeTransformation(start, ImmutableList.of(input));
+    } else {
```

Comment:
this logic is a little weird, it means a non-transform is passed into the TransformExecutor.  Think it would be better to take a list of start stages in the constructor rather than having special logic here.